docs/design-docs/06-config-as-data.md

```diff
@@ -0,0 +1,164 @@
+# Configuration as Data
+
+* Author(s): Martin Maly, @martinmaly
+* Approver: @bgrant0607
+
+## Why
+
+This document provides bacgrdound context for Package Orchestration, which is
+further elaborated in a dedicated [document](07-package-orchestration.md).
+
+## Configuration as Data
+
+*Configuration as Data* is an approach to management of configuration (incl.
+configuration of infrastructure, policy, services, applications, etc.) which:
+
+* makes configuration data the source of truth, stored separately from the live
+  state
+* uses a uniform, serializable data model to represent configuration
+* separates code that acts on the configuration from the data and from packages
+  / bundles of the data
+* abstracts configuration file structure and storage from operations that act
+  upon the configuration data; clients manipulating configuration data don’t
+  need to directly interact with storage (git, container images)
+
+![CaD Overview](./CaD%20Overview.svg)
+
+## Key Principles
+
+A system based on CaD *should* observe the following key principles:
+
+* secrets should be stored separately, in a secret-focused storage system
+  ([example](https://cloud.google.com/secret-manager))
+* stores a versioned history of configuration changes by change sets to bundles
+  of related configuration data
+* relies on uniformity and consistency of the configuration format, including
+  type metadata, to enable pattern-based operations on the configuration data,
+  along the lines of duck typing
+* separates schemas for the configuration data from the data, and relies on
+  schema information for strongly typed operations and to disambiguate data
+  structures and other variations within the model
+* decouples abstractions of configuration from collections of configuration data
+* represents abstractions of configuration generators as data with schemas, like
+  other configuration data
+* finds, filters / queries / selects, and/or validates configuration data that
+  can be operated on by given code (functions)
+* finds and/or filters / queries / selects code (functions) that can operate on
+  resource types contained within a body of configuration data
+* separates configuration data transformations from actuation, enabling
+  actuation by performing bulk changes to the configuration data; *actuation*,
```

Comment:
The partial definition of actuation is a bit hard to parse with this wording. Is "actuation" intended to be a separate bullet? Or "with actuation including..."?

Really this implies that reconciliation with the live state itself needs to be driven declaratively.

Reply:
Rewrote; LMK if this is better.